File: Sources/Core/TCController.cpp
/*
 *  TCContoller.h
 *
 *  Copyright 2010 Avérous Julien-Pierre
 *
 *  This file is part of TorChat.
 *
 *  TorChat is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  TorChat is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with TorChat.  If not, see <http://www.gnu.org/licenses/>.
 *
 */



#include <stdio.h>
#include <errno.h>

#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>

#include <string>
#include <map>

#include <Block.h>

#include "TCController.h"

#include "TCConfig.h"
#include "TCBuddy.h"
#include "TCTools.h"
#include "TCControlClient.h"



/*
** TCController - Constructor & Destructor
*/
#pragma mark -
#pragma mark TCController - Constructor & Destructor

TCController::TCController(TCConfig *_config) :
	config(_config),
	running(false)
{
	// Retain config
	_config->retain();
	
	// Init vars
	mstatus = tccontroller_available;
	running = false;
	socketAccept = 0;
	
	nQueue = 0;
	nBlock = NULL;
	
	buddiesLoaded = false;
	
	timer = 0;
	
	// Alloc queue
	mainQueue = dispatch_queue_create("com.torchat.core.controller.main", NULL);
	socketQueue = dispatch_queue_create("com.torchat.core.controller.socket", NULL);
}

TCController::~TCController()
{
	TCDebugLog("TCController Destructor");

	// Close client
	size_t i, cnt = clients.size();
		
	for (i = 0; i < cnt; i++)
	{
		clients[i]->stop();
		clients[i]->release();
	}
		
	clients.clear();
	
	
	// Stop buddies
	cnt = buddies.size();
		
	for (i = 0; i < cnt; i++)
	{
		buddies[i]->stop();
		buddies[i]->release();
	}
	buddies.clear();
		
	// Release delegate
	if (nBlock)
		Block_release(nBlock);
	nBlock = NULL;
		
	if (nQueue)
		dispatch_release(nQueue);
	nQueue = 0;
	
	// Release config
	config->release();

	// Release
	dispatch_release(mainQueue);
	dispatch_release(socketQueue);
}



/*
** TCController - Running
*/
#pragma mark -
#pragma mark TCController - Running

void TCController::start()
{
	dispatch_async_cpp(this, mainQueue, ^{
		
		if (running)
			return;
		
		if (!buddiesLoaded)
		{
			const tc_array	&sbuddies = config->buddies();
			size_t			i, cnt;
			
			//  -- Parse buddies --
			cnt = sbuddies.size();
			
			for (i = 0; i < cnt; i++)
			{
				tc_dictionary	item = sbuddies[i];
				TCBuddy			*buddy = new TCBuddy(config, item[TCConfigBuddyName], item[TCConfigBuddyAddress], item[TCConfigBuddyComment]);
				
				buddies.push_back(buddy);
				
				// Notify
				_notify(tcctrl_notify_buddy_new, "core_ctrl_note_new_buddy", buddy);
			}
			
			// -- Check that we are on buddy list --
			bool				found = false;
			const std::string	&self = config->get_self_address();
			
			cnt = buddies.size();
			
			for (i = 0; i < cnt; i++)
			{
				TCBuddy	*buddy = buddies[i];
				
				if (buddy->address().compare(self) == 0)
				{
					found = true;
					break;
				}
			}
			
			if (!found)
				addBuddy(config->localized("core_ctrl_myself"), self);
			
			// -- Buddy are loaded --
			buddiesLoaded = true;
		}
		
		// -- Start command server -- 
		struct sockaddr_in	my_addr;
		int					yes = 1;
				
		// > Configure the port and address
		my_addr.sin_family = AF_INET;
		my_addr.sin_port = htons(config->get_client_port());
		my_addr.sin_addr.s_addr = INADDR_ANY;
		memset(&(my_addr.sin_zero), '\0', 8);
				
		// > Instanciate the listening socket
		if ((sock = socket(AF_INET, SOCK_STREAM, 0)) == -1)
		{
			_error(tcctrl_error_serv_socket, "core_ctrl_err_socket", true);
			return;
		}
		
		// > Reuse the port
		if (setsockopt(sock, SOL_SOCKET, SO_REUSEADDR, &yes, sizeof(int)) == -1)
		{
			_error(tcctrl_error_serv_socket, "core_ctrl_err_setsockopt", true);
			return;
		}
		
		// > Bind the socket to the configuration perviously set
		if (bind(sock, (struct sockaddr *)&my_addr, sizeof(struct sockaddr)) == -1)
		{
			_error(tcctrl_error_serv_socket, "core_ctrl_err_bind", true);
			return;	
		}
		
		// > Set the socket as a listening socket
		if (listen(sock, 10) == -1)
		{
			_error(tcctrl_error_serv_socket, "core_ctrl_err_listen", true);
			return;	
		}
		
		// > Build a source
		socketAccept = dispatch_source_create(DISPATCH_SOURCE_TYPE_READ, sock, 0, socketQueue);
		
		// > Set the read handler
		dispatch_source_set_event_handler_cpp(this, socketAccept, ^{
			
			unsigned int		sin_size = sizeof(struct sockaddr);
			struct sockaddr_in	their_addr;
			int					csock;
			
			csock = accept(sock, (struct sockaddr *)&their_addr, &sin_size);
			
			if (csock == -1)
			{
				dispatch_async_cpp(this, mainQueue, ^{
					_error(tcctrl_error_serv_accept, "core_ctrl_err_accept", true);
				});
			}
			else
			{
				// Make the client async
				if (!doAsyncSocket(csock))
				{
					dispatch_async_cpp(this, mainQueue, ^{
						_error(tcctrl_error_serv_accept, "core_ctrl_err_async", true);
					});
					
					return;
				}

				// Add it later
				dispatch_async_cpp(this, socketQueue, ^{
					_addClient(csock);
				});
			}
		});
		
		// > Set the cancel handler
		dispatch_source_set_cancel_handler_cpp(this, socketAccept, ^{
			close(sock);
			sock = -1;
		});

		dispatch_resume(socketAccept);
		
		
		// -- Build a timer to keep alive buddies --
		timer = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, mainQueue);
		
		// Each 10s
		dispatch_source_set_timer(timer, DISPATCH_TIME_NOW, 10000000000L, 0);
		dispatch_source_set_event_handler_cpp(this, timer, ^{
			
			// Do nothing if not running
			if (!running || !buddiesLoaded)
				return;
			
			// (Re)start buddy (start do nothing if already started)
			size_t i, cnt = cnt = buddies.size();
			
			for (i = 0; i < cnt; i++)
				buddies[i]->start();
			
		});
		dispatch_resume(timer);
		
		// -- Start buddies --
		int i, cnt = buddies.size();
		
		for (i = 0; i < cnt; i++)
		{
			TCBuddy	*buddy = buddies[i];
			
			buddy->start();
		}
		
		// Give the status
		setStatus(mstatus);
		
		// Notify
		_notify(tcctrl_notify_started, "core_ctrl_note_started");
		
		// We are running !
		running = true;
	});
}

void TCController::stop()
{
	dispatch_async_cpp(this, mainQueue, ^{
		
		// Check if we are running
		if (!running)
			return;
		
		// Cancel the socket
		dispatch_source_cancel(socketAccept);
		dispatch_release(socketAccept);
		
		// Cancel the timer
		if (timer)
		{
			dispatch_source_cancel(timer);
			dispatch_release(timer);
		}
		
		socketAccept = 0;
		
		// Stop & release clients
		size_t i, cnt = clients.size();
		
		for (i = 0; i < cnt; i++)
		{
			clients[i]->stop();
			clients[i]->release();
		}
		clients.clear();
		
		// Stop buddies
		cnt = buddies.size();
		
		for (i = 0; i < cnt; i++)
			buddies[i]->stop();
				
		// Notify
		_notify(tcctrl_notify_stoped, "core_ctrl_note_stoped");
		
		running = false;
	});
}



/*
** TCController - Delegate
*/
#pragma mark -
#pragma mark TCController - Delegate

void TCController::setDelegate(dispatch_queue_t queue, tcctrl_event event)
{
	tcctrl_event cpy = NULL;
	
	if (event)
		cpy = Block_copy(event);
	
	if (queue)
		dispatch_retain(queue);
	
	// Asign on a block
	dispatch_async_cpp(this, mainQueue, ^{
		
		// Queue
		if (nQueue)
			dispatch_release(nQueue);
		
		nQueue = queue;
		
		
		// Block
		if (nBlock)
			Block_release(nBlock);
		nBlock = cpy;
	});
}



/*
** TCController - Status
*/
#pragma mark -
#pragma mark TCController - Status

void TCController::setStatus(tccontroller_status status)
{
	// Give the status
	dispatch_async_cpp(this, mainQueue, ^{
		
		// Hold internal status
		mstatus = status;
		
		// Run the controller if needed, else send status
		if (!running)
			start();
		else
		{
			// Give this status to buddy list
			size_t i, cnt = cnt = buddies.size();
			
			for (i = 0; i < cnt; i++)
			{
				TCBuddy	*buddy = buddies[i];
				
				buddy->sendStatus(status);
			}
		}
	});
}

tccontroller_status	TCController::status()
{
	__block tccontroller_status result = tccontroller_available;
	
	dispatch_sync_cpp(this, mainQueue, ^{
		result = mstatus;
	});
	
	return result;
}



/*
** TCController - Buddies
*/
#pragma mark -
#pragma mark TCController - Buddies

void TCController::addBuddy(const std::string &name, const std::string &address)
{
	addBuddy(name, address, "");
}

void TCController::addBuddy(const std::string &name, const std::string &address, const std::string &comment)
{
	TCBuddy		*buddy = new TCBuddy(config, name, address, comment);
	std::string	*cname = new std::string(name.c_str());
	std::string	*caddress = new std::string(address.c_str());
	std::string	*ccomment = new std::string(comment.c_str());
	
    dispatch_async_cpp(this, mainQueue, ^{
        
        // Add to the buddy list
        buddies.push_back(buddy);
		
		// Notify
		_notify(tcctrl_notify_buddy_new, "core_ctrl_note_new_buddy", buddy);
		
        // Start it
        buddy->start();
		
		// Save to config
		config->add_buddy(*caddress, *cname, *ccomment);
		
		// Clean
		delete cname;
		delete caddress;
		delete ccomment;
    });
}

void TCController::removeBuddy(const std::string &address)
{
	std::string *cpy = new std::string(address);
	
	dispatch_async_cpp(this, mainQueue, ^{
		
		size_t	i, cnt = buddies.size();
		
		// Search the buddy
		for (i = 0; i < cnt; i++)
		{
			if (buddies[i]->address().compare(*cpy) == 0)
			{
				// Stop and release
				TCBuddy	*buddy = buddies[i];
				
				buddy->stop();
				buddy->release();
				
				buddies.erase(buddies.begin() + i);
				
				// Save to config
				config->remove_buddy(*cpy);
				
				break;
			}
		}
				
		delete cpy;
	});
}

TCBuddy * TCController::getBuddyAddress(const std::string &address)
{    
	std::string		*addr = new std::string(address);
    __block TCBuddy *result = NULL;
	
	dispatch_sync_cpp(this, mainQueue, ^{
        
        size_t i, cnt = cnt = buddies.size();
		
		for (i = 0; i < cnt; i++)
		{
			TCBuddy	*buddy = buddies[i];
            
            if (buddy->address().compare(*addr) == 0)
            {
                result = buddy;
				result->retain();
				
                break;
            }
        }
		
		delete addr;
    });
	
    return result;
}

TCBuddy * TCController::getBuddyRandom(const std::string &random)
{
	std::string		*ran = new std::string(random);
    __block TCBuddy *result = NULL;
	
	dispatch_sync_cpp(this, mainQueue, ^{
        
        size_t i, cnt = cnt = buddies.size();
		
		for (i = 0; i < cnt; i++)
		{
			TCBuddy	*buddy = buddies[i];
            
            if (buddy->brandom().compare(*ran) == 0)
            {
                result = buddy;
				result->retain();
				
                break;
            }
        }
		
		delete ran;
    });
    
    return result;
}



/*
** TCController - TCControlClient
*/
#pragma mark -
#pragma mark TCController - TCControlClient

void TCController::cc_error(TCControlClient *client, TCInfo *serr)
{
	if (!client || !serr)
		return;
	
	
	// Give the error
	serr->retain();
	
	dispatch_async_cpp(this, mainQueue, ^{
		_send_event(serr);
		
		serr->release();
	});
	
	
	// Remove the client
	client->retain();

	dispatch_async_cpp(this, socketQueue, ^{
		
		std::vector<TCControlClient *>::iterator it;
		
		for (it = clients.begin(); it != clients.end(); it++)
		{
			TCControlClient *item = *it;
			
			if (item == client)
			{
				clients.erase(it);
				
				item->stop();
				item->release();
				
				break;
			}
		}
		
		// Realease it
		client->release();
	});
}

void TCController::cc_notify(TCControlClient *client, TCInfo *info)
{	
	if (!client || !info)
		return;
	
	info->retain();
	
	dispatch_async_cpp(this, mainQueue, ^{
		_send_event(info);
		
		info->release();
	});
}



/*
** TCController - Tools
*/
#pragma mark -
#pragma mark TCController - Tools

void TCController::_addClient(int sock)
{
	// > socketQueue <
	
	TCControlClient *client = new TCControlClient(config, sock);
	
	clients.push_back(client);
	
	client->start(this);
}



/*
** TCController - Helpers
*/
#pragma mark -
#pragma mark TCController - Helpers

void TCController::_error(tcctrl_info code, const std::string &info, bool fatal)
{
	// > mainQueue <
	
	TCInfo *err = new TCInfo(tcinfo_error, code, config->localized(info));
	
	_send_event(err);
	
	err->release();
	
	if (fatal)
		stop();
}

void TCController::_error(tcctrl_info code, const std::string &info, TCObject *ctx, bool fatal)
{
	// > mainQueue <
	
	TCInfo *err = new TCInfo(tcinfo_error, code, config->localized(info), ctx);
	
	_send_event(err);
	
	err->release();
	
	if (fatal)
		stop();
}

void TCController::_notify(tcctrl_info notice, const std::string &info)
{
	// > mainQueue <
	
	TCInfo *ifo = new TCInfo(tcinfo_info, notice, config->localized(info));
	
	_send_event(ifo);
	
	ifo->release();
}

void TCController::_notify(tcctrl_info notice, const std::string &info, TCObject *ctx)
{
	// > mainQueue <
	
	TCInfo *ifo = new TCInfo(tcinfo_info, notice, config->localized(info), ctx);
	
	_send_event(ifo);
	
	ifo->release();
}

void TCController::_send_event(TCInfo *info)
{
	// > mainQueue <
	
	if (!info)
		return;
	
	if (nQueue && nBlock)
	{
		info->retain();
		
		dispatch_async_cpp(this, nQueue, ^{
			
			nBlock(this, info);
			
			info->release();
		});
	}
}

File: TorChat/Sources/Core/TCController.m
/*
 *  TCContoller.h
 *
 *  Copyright 2013 Avérous Julien-Pierre
 *
 *  This file is part of TorChat.
 *
 *  TorChat is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  TorChat is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with TorChat.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>

#import "TCController.h"

#import "TCConfig.h"
#import "TCImage.h"

#import "TCControlClient.h"
#import "TCBuddy.h"
#import "TCBuddy.h"
#import "TCTools.h"



/*
** TCController
*/
#pragma mark - TCController

@interface TCController ()
{
	// -- Vars --
	// > Main Queue
	dispatch_queue_t		_localQueue;
	
	// > Timer
	dispatch_source_t		_timer;
	
	// > Accept Socket
	dispatch_queue_t		_socketQueue;
	dispatch_source_t		_socketAccept;
	int						_sock;
	
	// > Buddies
	BOOL					_buddiesLoaded;
	NSMutableArray			*_buddies;
	
	// > Config
	id <TCConfig>			_config;
	
	// > Clients
	NSMutableArray			*_clients;
	
	// > Status
	bool					_running;
	tccontroller_status		_mstatus;
	
	// > Delegate
	dispatch_queue_t		_delegateQueue;
	
	// > Profile
	TCImage					*_profileAvatar;
	NSString				*_profileName;
	NSString				*_profileText;
}

// -- Helpers --
- (void)_addClient:(int)sock;
- (void)_checkBlocked:(TCBuddy *)buddy;

- (void)_error:(tcctrl_info)code info:(NSString *)info fatal:(BOOL)fatal;
- (void)_error:(tcctrl_info)code info:(NSString *)info context:(id)ctx fatal:(BOOL)fatal;

- (void)_notify:(tcctrl_info)notice info:(NSString *)info;
- (void)_notify:(tcctrl_info)notice info:(NSString *)info context:(id)ctx;

- (void)_sendEvent:(TCInfo *)info;

@end



/*
** TCController
*/
#pragma mark - TCController

@implementation TCController


/*
** TCController - Instance
*/
#pragma mark - TCController - Instance

- (id)initWithConfiguration:(id <TCConfig>)config
{
	self = [super init];
	
	if (self)
	{
		config = _config;
		
		// Init vars
		_mstatus = tccontroller_available;

		// Get profile avatar
		_profileAvatar = [config profileAvatar];
		
		if (!_profileAvatar)
			_profileAvatar = [[TCImage alloc] initWithWidth:64 andHeight:64];
		
		// Get profile name & text
		_profileName = [config profileName];
		_profileText = [config profileText];
		
		// Alloc queue
		_localQueue = dispatch_queue_create("com.torchat.core.controller.local", DISPATCH_QUEUE_SERIAL);
		_socketQueue = dispatch_queue_create("com.torchat.core.controller.socket", DISPATCH_QUEUE_SERIAL);
		_delegateQueue = dispatch_queue_create("com.torchat.core.controller.delegate", DISPATCH_QUEUE_SERIAL);
		
		// Containers.
		_clients = [[NSMutableArray alloc] init];
		_buddies = [[NSMutableArray alloc] init];
	}
	
	return self;
}

- (void)dealloc
{
	TCDebugLog("TCController Destructor");
	
	// Close client
	for (TCControlClient *client in _clients)
		[client stop];

	// Stop buddies
	for (TCBuddy *buddy in _buddies)
		[buddy stop];
}



/*
** TCController - Life
*/
#pragma mark - TCController - Life

- (void)start
{
	dispatch_async(_localQueue, ^{
		
		if (_running)
			return;
		
		if (!_buddiesLoaded)
		{
			NSArray *sbuddies = [_config buddies];
			size_t	i, cnt;
			
			//  -- Parse buddies --
			cnt = [sbuddies count];
			
			for (i = 0; i < cnt; i++)
			{
				NSDictionary	*item = sbuddies[i];
				TCBuddy			*buddy = [[TCBuddy alloc] initWithConfiguration:_config alias:item[TCConfigBuddyAlias] address:item[TCConfigBuddyAddress] notes:item[TCConfigBuddyNotes]];
								
				// Check blocked status
				[self _checkBlocked:buddy];
				
				// Add to list
				[_buddies addObject:buddy];
				
				// Notify
				[self _notify:tcctrl_notify_buddy_new info:@"core_ctrl_note_new_buddy" context:buddy];
			}
			
			// -- Check that we are on the buddy list --
			BOOL		found = NO;
			NSString	*selfAddress = [_config selfAddress];
			
			
			for (TCBuddy	*buddy in _buddies)
			{
				if ([[buddy address] isEqualToString:selfAddress])
				{
					found = true;
					break;
				}
			}
			
			if (!found)
				[self addBuddy:[_config localized:@"core_ctrl_myself"] address:selfAddress];
			
			// -- Buddy are loaded --
			_buddiesLoaded = true;
		}
		
		// -- Start command server --
		struct sockaddr_in	my_addr;
		int					yes = 1;
		
		// > Configure the port and address
		my_addr.sin_family = AF_INET;
		my_addr.sin_port = htons([_config clientPort]);
		my_addr.sin_addr.s_addr = INADDR_ANY;
		memset(&(my_addr.sin_zero), '\0', 8);
		
		// > Instanciate the listening socket
		if ((_sock = socket(AF_INET, SOCK_STREAM, 0)) == -1)
		{
			[self _error:tcctrl_error_serv_socket info:@"core_ctrl_err_socket" fatal:YES];
			return;
		}
		
		// > Reuse the port
		if (setsockopt(_sock, SOL_SOCKET, SO_REUSEADDR, &yes, sizeof(int)) == -1)
		{
			[self _error:tcctrl_error_serv_socket info:@"core_ctrl_err_setsockopt" fatal:YES];
			return;
		}
		
		// > Bind the socket to the configuration perviously set
		if (bind(_sock, (struct sockaddr *)&my_addr, sizeof(struct sockaddr)) == -1)
		{
			[self _error:tcctrl_error_serv_socket info:@"core_ctrl_err_bind" fatal:YES];
			return;
		}
		
		// > Set the socket as a listening socket
		if (listen(_sock, 10) == -1)
		{
			[self _error:tcctrl_error_serv_socket info:@"core_ctrl_err_listen" fatal:YES];
			return;
		}
		
		// > Build a source
		_socketAccept = dispatch_source_create(DISPATCH_SOURCE_TYPE_READ, (uintptr_t)_sock, 0, _socketQueue);
		
		// > Set the read handler
		dispatch_source_set_event_handler(_socketAccept, ^{
			
			unsigned int		sin_size = sizeof(struct sockaddr);
			struct sockaddr_in	their_addr;
			int					csock;
			
			csock = accept(_sock, (struct sockaddr *)&their_addr, &sin_size);
			
			if (csock == -1)
			{
				dispatch_async(_localQueue, ^{
					[self _error:tcctrl_error_serv_accept info:@"core_ctrl_err_accept" fatal:YES];
				});
			}
			else
			{
				// Make the client async
				if (!doAsyncSocket(csock))
				{
					dispatch_async(_localQueue, ^{
						[self _error:tcctrl_error_serv_accept info:@"core_ctrl_err_async" fatal:YES];
					});
					
					return;
				}
				
				// Add it later
				dispatch_async(_socketQueue, ^{
					[self _addClient:csock];
				});
			}
		});
		
		// > Set the cancel handler
		dispatch_source_set_cancel_handler(_socketAccept, ^{
			close(_sock);
			_sock = -1;
		});
		
		dispatch_resume(_socketAccept);
		
		
		// -- Build a timer to keep alive buddies (start or sendStatus) --
		_timer = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, _localQueue);
		
		// Each 120s
		dispatch_source_set_timer(_timer, DISPATCH_TIME_NOW, 120000000000L, 0);
		dispatch_source_set_event_handler(_timer, ^{
			
			// Do nothing if not running
			if (!_running || !_buddiesLoaded)
				return;
			
			// (Re)start buddy (start do nothing if already started)
			for (TCBuddy *buddy in _buddies)
				[buddy keepAlive];
			
		});
		dispatch_resume(_timer);
		
		// -- Start buddies --
		for (TCBuddy *buddy in _buddies)
			[buddy start];
		
		// Give the status
		[self setStatus:_mstatus];
		
		// Notify
		[self _notify:tcctrl_notify_started info:@"core_ctrl_note_started"];
		
		// We are running !
		_running = YES;
	});
}

- (void)stop
{
	dispatch_async(_localQueue, ^{
		
		// Check if we are running
		if (!_running)
			return;
		
		// Cancel the socket
		dispatch_source_cancel(_socketAccept);
		
		// Cancel the timer
		if (_timer)
			dispatch_source_cancel(_timer);
		
		_socketAccept = nil;
		
		// Stop & release clients
		for (TCControlClient *client in _clients)
			[client stop];
		
		[_clients removeAllObjects];
		
		// Stop buddies
		for (TCBuddy *buddy in _buddies)
			[buddy stop];
		
		// Notify
		[self _notify:tcctrl_notify_stoped info:@"core_ctrl_note_stoped"];
		
		_running = false;
	});
}

// -- Status --
- (void)setStatus:(tccontroller_status)status
{
	// Give the status
	dispatch_async(_localQueue, ^{
		
		// Notify
		if (status != _mstatus)
			[self _notify:tcctrl_notify_status info:@"" context:@(status)];
		
		// Hold internal status
		_mstatus = status;
		
		// Run the controller if needed, else send status
		if (!_running)
			[self start];
		else
		{
			// Give this status to buddy list
			for (TCBuddy *buddy in _buddies)
				[buddy sendStatus:status];
		}
	});
}

- (tccontroller_status)status
{
	__block tccontroller_status result = tccontroller_available;
	
	dispatch_sync(_localQueue, ^{
		result = _mstatus;
	});
	
	return result;
}

// -- Profile --
- (void)setProfileAvatar:(TCImage *)avatar
{
	if (!avatar)
		return;
	
	// Set the avatar
	dispatch_async(_localQueue, ^{
		
		_profileAvatar = avatar;
		
		// Store avatar
		[_config setProfileAvatar:_profileAvatar];
		
		// Give this avatar to buddy list
		for (TCBuddy *buddy in _buddies)
			[buddy sendAvatar:_profileAvatar];
		
		// Notify
		[self _notify:tcctrl_notify_profile_avatar info:@"core_ctrl_note_profile_avatar" context:_profileAvatar];
	});
}

- (TCImage *)profileAvatar
{
	__block TCImage *result = NULL;
	
	dispatch_sync(_localQueue, ^{
		
		if (_profileAvatar)
			result = [_profileAvatar copy];
	});
	
	return result;
}


- (void)setProfileName:(NSString *)name
{
	if (!name)
		return;
	
	// Set the avatar
	dispatch_async(_localQueue, ^{
		
		// Hold the name
		_profileName = name;
		
		// Store the name
		[_config setProfileName:name];
		
		// Give this name to buddy list
		for (TCBuddy *buddy in _buddies)
			[buddy sendProfileName:_profileName];
		
		// Notify
		[self _notify:tcctrl_notify_profile_name info:@"core_ctrl_note_profile_name" context:_profileName];
	});
}

- (NSString *)profileName
{
	__block NSString *result = NULL;
	
	dispatch_sync(_localQueue, ^{
		result = _profileName;
	});
	
	return result;
}

- (void)setProfileText:(NSString *)text
{
	if (!text)
		return;
	
	// Set the avatar
	dispatch_async(_localQueue, ^{
		
		// Hold the text
		_profileText = text;
		
		// Store the text
		[_config setProfileText:text];
		
		// Give this text to buddy list
		for (TCBuddy *buddy in _buddies)
			[buddy sendProfileText:_profileText];

		// Notify
		[self _notify:tcctrl_notify_profile_text info:@"core_ctrl_note_profile_name" context:_profileText];
	});
}

- (NSString *)profileText
{
	__block NSString *result = NULL;
	
	dispatch_sync(_localQueue, ^{
		result = _profileText;
	});
	
	return result;
}

// -- Buddies --
- (void)addBuddy:(NSString *)name address:(NSString *)address
{
	[self addBuddy:name address:address comment:@""];
}

- (void)addBuddy:(NSString *)name address:(NSString *)address comment:(NSString *)comment
{
	if (!address)
		return;
	
	if (!name)
		name = @"";
	
	if (!comment)
		comment = @"";
	
	TCBuddy *buddy = [[TCBuddy alloc] initWithConfiguration:_config alias:name address:address notes:comment];
	
    dispatch_async(_localQueue, ^{
        
		// Check blocked status
		[self _checkBlocked:buddy];
		
        // Add to the buddy list
		[_buddies addObject:buddy];
		
		// Notify
		[self _notify:tcctrl_notify_buddy_new info:@"core_ctrl_note_new_buddy" context:buddy];
		
        // Start it
		[buddy start];
		
		// Save to config
		[_config addBuddy:address alias:name notes:comment];
    });
}

- (void)removeBuddy:(NSString *)address
{
	if (!address)
		return;
	
	dispatch_async(_localQueue, ^{
		
		NSUInteger	i, cnt = [_buddies count];
		
		// Search the buddy
		for (i = 0; i < cnt; i++)
		{
			TCBuddy *buddy = _buddies[i];
			
			if ([[buddy address] isEqualToString:address])
			{
				// Stop and release
				[buddy stop];
				
				[_buddies removeObjectAtIndex:i];
				
				// Save to config
				[_config removeBuddy:address];
				
				break;
			}
		}
	});
}

- (TCBuddy *)buddyWithAddress:(NSString *)address
{
	if (!address)
		return nil;
	
    __block TCBuddy *result = NULL;
	
	dispatch_sync(_localQueue, ^{
        
		for (TCBuddy *buddy in _buddies)
		{
			if ([[buddy address] isEqualToString:address])
			{
				result = buddy;
				break;
			}
        }
    });
	
    return result;
}

- (TCBuddy *)buddyWithRandom:(NSString *)random
{
	if (!random)
		return nil;
	
    __block TCBuddy *result = NULL;
	
	dispatch_sync(_localQueue, ^{
		
		for (TCBuddy *buddy in _buddies)
		{
			if ([[buddy random] isEqualToString:random])
            {
                result = buddy;
                break;
            }
        }
    });
    
    return result;
}

// -- Blocked Buddies --
- (BOOL)addBlockedBuddy:(NSString *)address
{
	__block BOOL result = false;
	
	dispatch_sync(_localQueue, ^{
		
		// Add the address to the configuration
		if ([_config addBlockedBuddy:address] == YES)
			result = YES;
	});
	
	// Mark the buddy as blocked
	if (result)
	{
		TCBuddy *buddy = [self buddyWithAddress:address];
		
		[buddy setBlocked:YES];
	}
	
	return result;
}

- (BOOL)removeBlockedBuddy:(NSString *)address
{
	__block BOOL result = false;
	
	dispatch_sync(_localQueue, ^{
		
		// Remove the address from the configuration
		if ([_config removeBlockedBuddy:address] == YES)
			result = YES;
	});
	
	// Mark the buddy as un-blocked
	if (result)
	{
		TCBuddy *buddy = [self buddyWithAddress:address];
		
		[buddy setBlocked:NO];
	}
	
	return result;
}

// -- TCControlClient --
- (void)cc_error:(TCControlClient *)client info:(TCInfo *)info
{
#warning Why not a delegate ?
	if (!client || !info)
		return;
	
	// Give the error
	dispatch_async(_localQueue, ^{
		[self _sendEvent:info];
	});
	
	// Remove the client
	dispatch_async(_socketQueue, ^{
		
		NSUInteger i, cnt = [_clients count];
		
		for (i = 0; i < cnt; i++)
		{
			TCControlClient *item = _clients[i];
			
			if (item == client)
			{
				[_clients removeObjectAtIndex:i];
				
				[item stop];
				break;
			}
		}
	});
}

- (void)cc_notify:(TCControlClient *)client info:(TCInfo *)info
{
#warning Why not a delegate ?

	if (!client || !info)
		return;
	
	dispatch_async(_localQueue, ^{
		[self _sendEvent:info];
	});
}

// -- Helpers --
- (void)_addClient:(int)csock
{
	// > socketQueue <
	
	TCControlClient *client = [[TCControlClient alloc] initWithConfiguration:_config andSocket:csock];
	
	[_clients addObject:client];
	
	[client startWithController:self];
}

- (void)_checkBlocked:(TCBuddy *)buddy
{
	// > localQueue <
	
	if (!_config)
		return;
	
	// XXX not thread safe
	NSArray	*blocked = [_config blockedBuddies];
	size_t	i, cnt = [blocked count];
	
	[buddy setBlocked:NO];
	
	// Search
	for (i = 0; i < cnt; i++)
	{
		NSString *address = blocked[i];
		
		if ([address isEqualToString:[buddy address]])
		{
			[buddy setBlocked:YES];
			[buddy stop];
			break;
		}
	}
}


- (void)_error:(tcctrl_info)code info:(NSString *)info fatal:(BOOL)fatal
{
	// > localQueue <
	
	TCInfo *err = [TCInfo infoOfKind:tcinfo_error infoCode:code infoString:[_config localized:info]];

	[self _sendEvent:err];
		
	if (fatal)
		[self stop];
}

- (void)_error:(tcctrl_info)code info:(NSString *)info context:(id)ctx fatal:(BOOL)fatal
{
	// > localQueue <
		
	TCInfo *err = [TCInfo infoOfKind:tcinfo_error infoCode:code infoString:[_config localized:info] context:ctx];
	
	[self _sendEvent:err];
	
	if (fatal)
		[self stop];
}

- (void)_notify:(tcctrl_info)notice info:(NSString *)info
{
	// > localQueue <
		
	TCInfo *ifo = [TCInfo infoOfKind:tcinfo_info infoCode:notice infoString:[_config localized:info]];

	[self _sendEvent:ifo];
}

- (void)_notify:(tcctrl_info)notice info:(NSString *)info context:(id)ctx
{
	// > localQueue <
	
	TCInfo *ifo = [TCInfo infoOfKind:tcinfo_info infoCode:notice infoString:[_config localized:info] context:ctx];

	[self _sendEvent:ifo];
}

- (void)_sendEvent:(TCInfo *)info
{
	// > localQueue <
	
	if (!info)
		return;
	
	id <TCControllerDelegate> delegate = self.delegate;
	
	if (delegate)
	{
		dispatch_async(_delegateQueue, ^{
			[delegate torchatController:self information:info];
		});
	}
}

@end
